[REF] Master simplify model startup vsc

VincentSchippefilt opened the pull request:
## Description:

preliminary work to split the model into core_model and all_the_rest_model

Task: : 3637435 (https://www.odoo.com/web#id=3637435&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

VincentSchippefilt committed:
[IMP] core_ui: core evaluation plugins

Before this commit, the core_ui_plugins where of type UIPlugin, which is not correct

After this commit, they have their own type: CoreUiPlugin, which is basically a core
plugin that has access to all getters

Task: 3637435

VincentSchippefilt committed:
[REF] base_plugin: simplify history update in base plugin

Simplify the creation and binding on the history object.

This is preliminary work for separating Model.ts into multiple
independant parts.

Task: 3637435

VincentSchippefilt commented:
robodoo rebase-ff

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/4160).

robodoo commented:
Merge method set to rebase and fast-forward.

hokolomopo commented on src/plugins/ui_plugin.ts:
```diff
@@ -27,7 +27,7 @@ export interface UIPluginConfig {
   readonly session: Session;
   readonly defaultCurrencyFormat?: Format;
-  readonly customColors: Color[];
+  // readonly customColors: Color[];
```
comment ?

hokolomopo commented on src/model.ts:
```diff
@@ -249,8 +247,7 @@ export class Model extends EventBus<any> implements CommandDispatcher {
 
     for (let Plugin of coreViewsPluginRegistry.getAll()) {
-      const plugin = this.setupUiPlugin(Plugin);
-      this.coreViewsPlugins.push(plugin);
+      const plugin = this.setupCoreUiPlugin(Plugin);
```
I didn't think `coreViewPlugin` was a very good name, but IMO CoreUIPlugin is even more confusing.

Anyway the registry should also be renamed if we want to change the nomenclature

hokolomopo commented on src/model.ts:
```diff
@@ -324,6 +321,20 @@ export class Model extends EventBus<any> implements CommandDispatcher {
 
+  private setupCoreUiPlugin(Plugin: CorePluginConstructor) {
+    const plugin = new Plugin({ ...this.corePluginConfig, getters: this.getters });
+    for (let name of Plugin.getters) {
+      if (!(name in plugin)) {
+        throw new Error(`Invalid getter name: ${name} for plugin ${plugin.constructor}`);
+      }
+      if (name in this.getters) {
+        throw new Error(`Getter "${name}" is already defined.`);
+      }
+      this.getters[name] = plugin[name].bind(plugin);
+    }
```
Any way to extract this loop in a method ? It's now copy/pasted 3 times in this file

hokolomopo reviewed (commented):
Tchou tchou :steam_locomotive: 

Not a fan of the name `CoreUIPlugin`, but I don't have ideas either :shrug:

LucasLefevre commented on src/plugins/ui_core_views/header_sizes_ui.ts:
```diff
@@ -21,7 +21,7 @@ interface CellWithSize {
   size: Pixel;
 }
 
-export class HeaderSizeUIPlugin extends UIPlugin<HeaderSizeState> implements HeaderSizeState {
```
note that this plugin depends on DOM api to create a canvas element. We won't be able to move it to a Worker as it is now, we'll have to refactor a bit more here.
`private ctx = document.createElement("canvas").getContext("2d")`

LucasLefevre replied:
I didn't know OffscreenCanvas were a thing

LucasLefevre reviewed (commented):
> the core_ui_plugins where of type UIPlugin, which is not correct

could you elaborate in the commit message why it's not correct?
It's actually not clear to me why it's not correct. Core UI plugins still handle UI commands, can use all getters. Except what's available in the plugin config, nothing changed much.

They should probably not be able to handle UI commands, but then this commit is not enough (there's currently a few exceptions: START, EVALUATE_CELLS,  REFRESH_PIVOT, others?). 

Also note that moving the evaluation to a worker will be very hard, even impossible. It depends on a loooot of things, especially for odoo pivots (rpc, web pivot model, and more)

VincentSchippefilt commented on src/model.ts:
```diff
@@ -587,13 +598,13 @@ export class Model extends EventBus<any> implements CommandDispatcher {
     const isCommandCore = isCoreCommand(command);
     for (const handler of handlers) {
-      if (!isCommandCore && handler instanceof CorePlugin) {
+      if (!isCommandCore && this.corePlugins.includes(handler as any)) {
```
yes, I'm looking into it.

VincentSchippefilt committed:
[DOC] history: comments

Task: 3637435

VincentSchippefilt committed:
[REF] model: dispatch to separate groups of plugins

separating the different groups of plugins have 2 advantages and 1 disadvantage:
+ it allows when reading the code of model to see to whom you are dispatching, from the different parts of the code
+ it reduced slightly the computing time, because we only dispatch to the correct plugins
- it requires to duplicate code a bunch

Task: 3637435

VincentSchippefilt committed:
[REF] model: recordChanges of core commands

when dispatching a root command, record the changes only of the core plugins

Task: 3637435

VincentSchippefilt committed:
[REF] collaborative: building collaborative session

pass multiple arguments instead of 1 objects with keys
Adds a few comments in the different dispatch methods to follow callers

Task: 3637435

VincentSchippefilt committed:
[REF] model: remove unused functions

Task: 3637435
